File: cs/OpenAxis/Geometry/PoseHelpers.cs
using System;

namespace OpenAxis.Geometry
{
    /// <summary>
    /// Protocol-neutral camera pose value used by the geometry conversion helpers.
    /// Position and rotation are expressed in the caller's world coordinates.
    /// </summary>
    public readonly struct CameraPoseValue
    {
        public Vec3 Position { get; }
        public Vec3 RotationVector { get; }
        public double? Fov { get; }
        public double? OrthoExtent { get; }

        public CameraPoseValue(
            Vec3 position,
            Vec3 rotationVector,
            double? fov = null,
            double? orthoExtent = null)
        {
            if (fov.HasValue && orthoExtent.HasValue)
                throw new ArgumentException("A camera pose value cannot contain both perspective and orthographic projection");
            if (fov.HasValue && (!IsFinite(fov.Value) || fov.Value <= 0))
                throw new ArgumentOutOfRangeException(nameof(fov), "FOV must be positive and finite");
            if (orthoExtent.HasValue && (!IsFinite(orthoExtent.Value) || orthoExtent.Value <= 0))
                throw new ArgumentOutOfRangeException(nameof(orthoExtent), "Orthographic extent must be positive and finite");
            if (!IsFinite(position) || !IsFinite(rotationVector))
                throw new ArgumentException("Camera position and rotation vector must be finite");

            Position = position;
            RotationVector = rotationVector;
            Fov = fov;
            OrthoExtent = orthoExtent;
        }

        private static bool IsFinite(Vec3 value) =>
            IsFinite(value.X) && IsFinite(value.Y) && IsFinite(value.Z);

        private static bool IsFinite(double value) =>
            !double.IsNaN(value) && !double.IsInfinity(value);
    }

    /// <summary>
    /// Helpers for converting between eye/target/up camera representations
    /// and the OpenAxis wire format (position + rotation vector).
    ///
    /// The semantic camera frame is right-handed: +Z is backward (away from
    /// target), +Y is up, and +X is right. In a left-handed client world the
    /// pose rotation is the proper factor of that frame; world.orientation
    /// carries the reflection separately. These eye/target/up helpers produce
    /// that factor without requiring a handedness argument.
    /// </summary>
    public static class PoseHelpers
    {
        /// <summary>Convert eye/target/up camera to a position and rotation-vector pose.</summary>
        public static CameraPoseValue PoseFromLookAt(
            Vec3 eye, Vec3 target, Vec3 up,
            double? fov = null, double? orthoExtent = null)
        {
            var backward = (eye - target).Normalized();
            var distance = (eye - target).Length();
            if (distance < 1e-12)
            {
                return new CameraPoseValue(eye, new Vec3(0, 0, 0), fov, orthoExtent);
            }

            var right = up.Cross(backward).Normalized();
            if (right.Length() < 1e-6)
            {
                var fallback = Math.Abs(up.Y) > 0.9 ? new Vec3(1, 0, 0) : new Vec3(0, 1, 0);
                right = fallback.Cross(backward).Normalized();
            }
            var upOrtho = backward.Cross(right).Normalized();

            var q = Quat.FromAxes(right, upOrtho, backward).Normalize();
            var rv = q.ToRotvec();

            return new CameraPoseValue(
                eye,
                new Vec3(rv.wx, rv.wy, rv.wz),
                fov,
                orthoExtent);
        }

        /// <summary>
        /// Convert a position and rotation-vector pose to eye/target/up vectors.
        /// If pivot is given, computes target distance from (eye-pivot).backward.
        /// </summary>
        public static (Vec3 eye, Vec3 target, Vec3 up) LookAtFromPose(
            CameraPoseValue pose,
            double defaultDistance = 10.0,
            Vec3? pivot = null)
        {
            var rotation = pose.RotationVector;
            var q = Quat.FromRotvec(rotation.X, rotation.Y, rotation.Z);
            var eye = pose.Position;
            var up = q.Rotate(new Vec3(0, 1, 0));
            var backward = q.Rotate(new Vec3(0, 0, 1));

            double distance;
            if (pivot != null)
            {
                var depth = (eye - pivot.Value).Dot(backward);
                distance = Math.Max(depth, 0.01);
            }
            else
            {
                distance = defaultDistance;
            }

            var target = eye - backward * distance;
            return (eye, target, up);
        }
    }
}

File: ts/sdk/src/geometry/pose-helpers.ts
import { Quat } from "./quat.js";
import { Vec3 } from "./vec3.js";

/** Protocol-neutral camera pose value expressed in the caller's world coordinates. */
export interface CameraPoseValue {
  readonly position: Vec3;
  readonly rotationVector: Vec3;
  readonly fov?: number;
  readonly orthoExtent?: number;
}

export interface LookAtValue {
  readonly eye: Vec3;
  readonly target: Vec3;
  readonly up: Vec3;
}

function validatePose(position: Vec3, rotationVector: Vec3, fov?: number, orthoExtent?: number): CameraPoseValue {
  if (fov !== undefined && orthoExtent !== undefined) throw new RangeError("A camera pose cannot contain both projections");
  if (fov !== undefined && (!Number.isFinite(fov) || fov <= 0)) throw new RangeError("FOV must be positive and finite");
  if (orthoExtent !== undefined && (!Number.isFinite(orthoExtent) || orthoExtent <= 0)) throw new RangeError("Orthographic extent must be positive and finite");
  const components = [position.x, position.y, position.z, rotationVector.x, rotationVector.y, rotationVector.z];
  if (!components.every(Number.isFinite)) throw new RangeError("Camera position and rotation vector must be finite");
  return { position, rotationVector, fov, orthoExtent };
}

/** Convert eye/target/up to the OpenAxis camera-local rotation convention. */
export function poseFromLookAt(
  eye: Vec3,
  target: Vec3,
  up: Vec3,
  projection: { fov?: number; orthoExtent?: number } = {},
): CameraPoseValue {
  const offset = eye.sub(target);
  if (offset.length() < 1e-12) {
    return validatePose(eye, new Vec3(0, 0, 0), projection.fov, projection.orthoExtent);
  }
  const backward = offset.normalized();
  let right = up.cross(backward).normalized();
  if (right.length() < 1e-6) {
    const fallback = Math.abs(up.y) > 0.9 ? new Vec3(1, 0, 0) : new Vec3(0, 1, 0);
    right = fallback.cross(backward).normalized();
  }
  const orthogonalUp = backward.cross(right).normalized();
  const rotationVector = Vec3.fromArray(Quat.fromAxes(right, orthogonalUp, backward).normalize().toRotvec());
  return validatePose(eye, rotationVector, projection.fov, projection.orthoExtent);
}

/** Convert an OpenAxis camera pose to eye/target/up values. */
export function lookAtFromPose(
  pose: CameraPoseValue,
  defaultDistance = 10,
  pivot?: Vec3,
): LookAtValue {
  const q = Quat.fromRotvec(pose.rotationVector.x, pose.rotationVector.y, pose.rotationVector.z);
  const up = q.rotate(new Vec3(0, 1, 0));
  const backward = q.rotate(new Vec3(0, 0, 1));
  const distance = pivot === undefined
    ? defaultDistance
    : Math.max(pose.position.sub(pivot).dot(backward), 0.01);
  return { eye: pose.position, target: pose.position.sub(backward.mul(distance)), up };
}
